src/fmm.jl

```diff
@@ -0,0 +1,296 @@
+            val = rfspace(point.point)
+            for localbasis in eachindex(val)
+                for data in tad.data[:,localbasis,ncell]
+                    if data[1] != 0 && ind + npoint != 0
+                        rc[sind, 1] = ind + npoint
+                        rc[sind, 2] = data[1]
+                        vals[sind, 1] = val[localbasis].value[1] * point.weight * data[2]
+                        vals[sind, 2] = val[localbasis].value[2] * point.weight * data[2]
+                        vals[sind, 3] = val[localbasis].value[3] * point.weight * data[2]
+                        sind += 1
+                    end 
+                end 
+            end
+        end
+    end
+
+    return rc, vals
+
+end
+
+function getBmatrix_curl(qp::Matrix, X::BEAST.Space)
+    rfspace = refspace(X)
+    _, tad, _ = assemblydata(X)
+    len = length(qp) * length(qp[1,1]) * size(tad.data)[1] * size(tad.data)[2]
+    rc = ones(Int, len, 2)
+    vals = zeros(Float64, len, 3)
+    sind = 1
+
+    for (ncell, cell) in enumerate(qp[1,:])
+        ind = (ncell - 1) * length(cell)
+        for (npoint, point) in enumerate(cell)
+            val = rfspace(point.point)
+            for localbasis in eachindex(val)
+                for data in tad.data[:,localbasis,ncell]
+                    if data[1] != 0 && ind + npoint != 0
+                        rc[sind, 1] = ind + npoint
+                        rc[sind, 2] = data[1]
+                        vals[sind, 1] = val[localbasis].curl[1] * point.weight * data[2]
+                        vals[sind, 2] = val[localbasis].curl[2] * point.weight * data[2]
+                        vals[sind, 3] = val[localbasis].curl[3] * point.weight * data[2]
+                        sind += 1
+                    end 
+                end 
+            end
+        end
+    end
+
+    return rc, vals
+end
+
+function getBmatrix_div(qp::Matrix, X::BEAST.Space)
+    rfspace = refspace(X)
+    _, tad, _ = assemblydata(X)
+    len = length(qp) * length(qp[1,1]) * size(tad.data)[1] * size(tad.data)[2]
+    rc = ones(Int, len, 2)
+    vals = zeros(Float64, len)
+    sind = 1
+
+    for (ncell, cell) in enumerate(qp[1,:])
+        ind = (ncell - 1) * length(cell)
+        for (npoint, point) in enumerate(cell)
+            val = rfspace(point.point)
+            for localbasis in eachindex(val)
+                for data in tad.data[:,localbasis,ncell]
+                    if data[1] != 0 && ind + npoint != 0
+                        rc[sind, 1] = ind + npoint
+                        rc[sind, 2] = data[1]
+                        vals[sind] = val[localbasis].divergence * point.weight * data[2]
+                        sind += 1
+                    end 
+                end 
+            end
+        end
+    end
+
+    return rc, vals
+end
+
+function getnormals(qp::Matrix)
+    normals = zeros(Float64, length(qp)*length(qp[1,1]), 3)
+    for (i, points) in enumerate(qp[1,:])
+        for (j, point) in enumerate(qp[1,i])
```

Comment:
I am wondering if it could make sense to narrow down the element type of qp::Matrix. While there is reason not do be overly specific, it seems not to be a matrix containing numbers (otherwise an enumerate on qp[i,j] would not make sense, would it?)